criu/pages-compress.c

```diff
@@ -0,0 +1,237 @@
+#include <limits.h>
+#include <pthread.h>
+#include <sched.h>
+#include <stdio.h>
+#include <sys/resource.h>
+#include <unistd.h>
+
+#include <compel/plugins/std/syscall-codes.h>
+
+#include "pages-compress.h"
+
+#include "crtools.h"
+#include "image.h"
+#include "log.h"
+#include "memfd.h"
+#include "sched.h"
+#include "util.h"
+
+#include <lz4/programs/lz4io.h>
+#include <lz4/lib/lz4frame.h>
+
+#undef LOG_PREFIX
+#define LOG_PREFIX "decompress: "
+
+const char *temp_path_decompressed = "/tmp/decompressed";
+
+int compress_images(void)
+{
+	char format[PATH_MAX];
+	char pathsrc[PATH_MAX];
+	char pathdst[PATH_MAX];
+	const long pid = getpid();
+	const int dfd = get_service_fd(IMG_FD_OFF);
+	const int id = 1; // the only image we'd like to compress
+	int ret;
+
+	LZ4IO_prefs_t *lz4_prefs = LZ4IO_defaultPreferences();
+	LZ4IO_setBlockSize(lz4_prefs, 128 * 1024);
+
+	pr_debug("Compressing pages\n");
+
+	snprintf(format, sizeof(format), "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES].fmt);
+	snprintf(pathsrc, sizeof(pathsrc), format, pid, dfd, id);
+
+	snprintf(format, sizeof(format), "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES_COMP].fmt);
+	snprintf(pathdst, sizeof(pathdst), format, pid, dfd, id);
+
+	ret = LZ4IO_compressFilename(pathsrc, pathdst, 8, lz4_prefs);
+	LZ4IO_freePreferences(lz4_prefs);
+
+	if (ret) {
+		pr_err("Can't compress %s to %s\n", pathsrc, pathdst);
+		return -1;
+	}
+	if (unlink(pathsrc)) {
+		pr_perror("Can't delete uncompressed source pages image: %s\n", pathsrc);
+		return -1;
+	}
+	return 0;
+}
+
+static int decompress_image(int comp_fd) {
+
+	const int compbufsize = 64 * 1024;
+	char compbuf[compbufsize];
+	const int outbufsize = 4 * 1024 * 1024; // LZ4F_max4MB is the max block size; see LZ4F_blockSizeID_t for details
+	char outbuf[outbufsize];
+	LZ4F_errorCode_t lz4err;
+	LZ4F_decompressionContext_t dctx;
+	int decomp_fd;
+	size_t totalread = 0;
+	size_t totalwrite = 0;
+	int bytestoread = compbufsize;
+	size_t offset = 0;
+
+	lz4err = LZ4F_createDecompressionContext(&dctx, LZ4F_VERSION);
+	if (LZ4F_isError(lz4err)) {
+		pr_err("Can't create LZ4 decompression context\n");
+		return -1;
+	}
+
+	pr_debug("Creating file at '%s'\n", temp_path_decompressed);
+	decomp_fd = open(temp_path_decompressed, O_CREAT | O_WRONLY | O_TRUNC, S_IRUSR | S_IWUSR);
+
+	if (0 > decomp_fd) {
+		pr_err("Can't create file, errno=%d\n", errno);
+		LZ4F_freeDecompressionContext(dctx);
+		return -1;
+	}
+
+	while (1) {
+		const ssize_t readbytes = read(comp_fd, compbuf + offset, bytestoread);
+		if (0 > readbytes) {
+			if (EINTR == errno) {
+				continue;
+			}
+			pr_perror("failed to read compressed data");
+			break; // stop on error
+		}
+		if (!readbytes && 0 == offset) {
```

Comment:
I believe offset!=0 happens if the decompressing function needs more data to fully process it (such as with incompletely read header). If EOF comes in such moment it should be an error. Current code will AFAICS lock up for readbytes==0 && 0!=offset.

Reply:
I believe it's possible that there can be a tail after LZ4F_decompress call when there is no more data to read from file (e.g. it doesn't fit output buffer), so it should be decompressed the next LZ4F_decompress call.

Reply:
OK, I agree, I made a mistake in my review here.